ambari-server/src/main/java/org/apache/ambari/server/security/encryption/PasswordPropertiesEncryptor.java

```diff
@@ -0,0 +1,121 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ambari.server.security.encryption;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.ambari.server.AmbariException;
+import org.apache.ambari.server.AmbariRuntimeException;
+import org.apache.ambari.server.state.Cluster;
+import org.apache.ambari.server.state.Clusters;
+import org.apache.ambari.server.state.PropertyInfo.PropertyType;
+import org.apache.ambari.server.utils.TextEncoding;
+
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+/**
+ * {@link Encryptor} implementation for encrypting/decrypting PASSWORD type properties
+ *
+ */
+    
+@Singleton
+public class PasswordPropertiesEncryptor implements Encryptor<Map<String, String>> {
+
+  private static final String ENCRYPTED_PROPERTY_PREFIX = "${enc=aes256_hex, value=";
+  private static final String ENCRYPTED_PROPERTY_SCHEME = ENCRYPTED_PROPERTY_PREFIX + "%s}";
+
+  private final EncryptionService encryptionService;
+  private final Map<Long, Map<String, Set<String>>> clusterPasswordProperties = new HashMap<>();
+
+  @Inject
+  public PasswordPropertiesEncryptor(Clusters clusters, EncryptionService encryptionService) {
+    this.encryptionService = encryptionService;
+  }
+
+  @Override
+  public Map<String, String> encryptSensitiveData(Map<String, String> properties, Object... additionaKeys) {
+    try {
+      final Map<String, String> encryptedProperties = new HashMap<>(properties.size());
+      if (properties != null) {
+        final Cluster cluster = (Cluster) additionaKeys[0];
+        final String configType = (String) additionaKeys[1];
+        final Set<String> passwordProperties = getPasswordProperties(cluster, configType);
+        if (passwordProperties != null && !passwordProperties.isEmpty()) {
+          String encryptedValue;
+          for (Map.Entry<String, String> property : properties.entrySet()) {
+            encryptedValue = passwordProperties.contains(property.getKey()) && !isEncryptedPassword(property.getValue())
+                ? encryptAndDecoratePropertyValue(property.getValue())
+                : property.getValue();
+            encryptedProperties.put(property.getKey(), encryptedValue);
+          }
+        }
+      }
+      return encryptedProperties;
+    } catch (Exception e) {
+      throw new AmbariRuntimeException("Error while encrypting sensitive data", e);
+    }
+  }
+
+  private boolean isEncryptedPassword(String password) {
+    return password != null && password.startsWith(ENCRYPTED_PROPERTY_PREFIX); // assuming previous encryption by this class
+  }
+
+  private Set<String> getPasswordProperties(Cluster cluster, String configType) throws AmbariException {
+    final long clusterId = cluster.getClusterId();
```

Comment:
How will a stack upgrade affect this?

Reply:
nice catch; I'll handle that situation like this: subscribing to org.apache.ambari.server.events.StackUpgradeFinishEvent and clean the map when this happened.

Reply:
No objection here.  Thanks.

Reply:
During a stack upgrade, new properties are added and possibly properties change types.   How will this cache behave?

Reply:
Upon stack upgrade finish the cache will be purged. So that proceeding encryption actions will re-build it.
See the latest patch I submitted